lib/websocket/frame.js

```diff
@@ -0,0 +1,223 @@
+ */
+  constructor ({
+    opcode = opcodes.TEXT,
+    fin = false,
+    rsv1 = false,
+    rsv2 = false,
+    rsv3 = false,
+    payloadLength,
+    data
+  } = {}) {
+    this.fin = fin
+    this.rsv1 = rsv1
+    this.rsv2 = rsv2
+    this.rsv3 = rsv3
+    this.opcode = opcode
+
+    this.#buffers = [data]
+    this.#offset += data.length
+
+    this.payloadLength = payloadLength
+    this.fragmented = !this.fin && this.opcode !== opcodes.CONTINUATION
+  }
+
+  get data () {
+    return Buffer.concat(this.#buffers, this.#offset)
+  }
+
+  get dataOffset () {
+    return this.#offset
+  }
+
+  addFrame (buffer) {
+    this.#buffers.push(buffer)
+    this.#offset += buffer.length
+  }
+
+  parseCloseBody (onlyCode) {
+    // https://datatracker.ietf.org/doc/html/rfc6455#section-7.1.5
+    /** @type {number|undefined} */
+    let code
+
+    const { data } = this
+
+    if (data.length > 2) {
+      // _The WebSocket Connection Close Code_ is
+      // defined as the status code (Section 7.4) contained in the first Close
+      // control frame received by the application
+      code = data.readUInt16BE(0)
+    }
+
+    if (onlyCode) {
+      if (!isValidStatusCode(code)) {
+        return null
+      }
+
+      return { code }
+    }
+
+    // https://datatracker.ietf.org/doc/html/rfc6455#section-7.1.6
+    /** @type {Buffer} */
+    let reason = data.subarray(2)
+
+    // Remove BOM
+    if (reason[0] === 0xEF && reason[1] === 0xBB && reason[2] === 0xBF) {
+      reason = reason.subarray(3)
+    }
+
+    if (code !== undefined && !isValidStatusCode(code)) {
+      return null
+    }
+
+    try {
+      // TODO: optimize this
+      reason = new TextDecoder('utf-8', { fatal: true }).decode(reason)
+    } catch {
+      return null
+    }
+
+    return { code, reason }
+  }
+
+  /**
+   * @param {Buffer} buffer
+   */
+  static from (buffer) {
+    const fin = (buffer[0] & 0x80) !== 0
+    const rsv1 = (buffer[0] & 0x40) !== 0
+    const rsv2 = (buffer[0] & 0x20) !== 0
+    const rsv3 = (buffer[0] & 0x10) !== 0
+    const opcode = buffer[0] & 0x0F
+    const masked = (buffer[1] & 0x80) !== 0
+
+    // Data sent from an endpoint cannot be masked.
+    assert(!masked)
+
+    let payloadLength = buffer[1] & 0x7F
+    let data
+
+    if (payloadLength <= 125) {
+      data = buffer.subarray(2)
+    } else if (payloadLength === 126) {
+      payloadLength = buffer.subarray(2, 4).readUInt16BE(0)
+      data = buffer.subarray(4)
+    } else if (payloadLength === 127) {
+      payloadLength = buffer.subarray(2, 10).readBigUint64BE(0)
+      data = buffer.subarray(10)
+    }
+
+    const frame = new WebsocketFrame({
+      fin,
+      rsv1,
+      rsv2,
+      rsv3,
+      opcode,
+      payloadLength,
+      data
+    })
+
+    return frame
+  }
+}
+
+class WebsocketFrameSend {
+  /**
+   * @param {Buffer|undefined} data
+   */
+  constructor (data) {
+    this.frameData = data
+    this.maskKey = randomBytes(4)
```

Comment:
This can be slow and often unnecessary... is there a way we can force 0 mask?

Reply:
Unsure of what you mean - it's possible to create the key lazily, but it has to be secure.

> Clients MUST choose a new masking key for each frame, using an
   algorithm that cannot be predicted by end applications that provide
   data.  For example, each masking could be drawn from a
   cryptographically strong random number generator.

Reply:
The mask is not useful/necessary for server side websocket applications and a significant performance bottleneck. Without the possibility for the 0 mask optimization, this implementation would not be useful at least for us. The only time the mask is useful and the reason it exists is for when you non trusted proxies in the path.

Reply:
I see now. I'd need to give it more thought as it goes against everything the spec tells us to do. Does ws have a similar ability?

Reply:
Yes, but it uses `crypto.randomBytes()` by default.

Reply:
>> Does ws have a similar ability?

Yes, @lpinca was kind enough to add the option.

```js
new ws.WebSocket(url, {
   generateMask() {},
})
```

Reply:
This would be better off in its own PR so it can be fleshed out more. The spec WebSocket takes 2 arguments - neither of which is an object - so it's not as easy to add custom things like it is with fetch. Also considering how it violates the rfc I'd need to give it much more thought as mentioned above.

This PR is fine for implementing just a bare-bones version of WebSockets, so we can add in more stuff later.